### PR TITLE
[Docs Rewrite] Create the "Quick Start" page

### PR DESCRIPTION
**Pull request by markerikson**
## PR Type

**Does this PR add a _new_ page, or update an _existing_ page?**

Adds the new "Quick Start" page

## Checklist

- [x] Is there an existing issue for this PR?
  - #3674 
- [ ] Have the files been linted and formatted?

## What docs page is being added or updated?

- **Section**: Introduction
- **Page**: Quick Start

### What kind of content category is this page (tutorial, how-to, explanation, reference)?

The "Quick Start" page is a form of tutorial, but intended to demonstrate the "fastest" / "easiest" / "correct" way to use Redux without explaining _why_ things work this way.  

### Who is the intended target audience?

Beginners who know nothing about Redux, yet.

#### What knowledge are we assuming they have?

- JS syntax: ES6, `async/await`, object spreads
- React: components, hooks
- Other: HTTP / AJAX, `fetch`

### What are the intended results or takeaways from reading this page?

Readers should be able to mimic the demonstrated patterns to create a basic React+Redux app that does some CRUD work.

### What is the most critical info they should learn?

- Redux is a tool for managing state outside of your React components
- The Redux Toolkit package is the simplest and recommended way to use Redux
- Key React+Redux APIs (`configureStore`, `createSlice`, `useSelector` / `useDispatch`)
- Immutability is vital, but either requires hand-written updates or use of RTK+Immer "magic"

## Status

Hacked together this initial skeleton of a page.  So far I've done:

- Figured out how to use MDX with Docusaurus
- Added reusable components for `<Note>` (including multiple styles based on the Cypress.io docs) and `<DetailedExplanation>`
- Created an empty "Quick Start" page
- Added some initial "what is this and why would I use it?" content
- Added instructions on creating a project

I got stuck with some yak shaving when I realized that using the `diff` highlighting format leaves the `+-` diff signs as copyable characters, and ideally I'd like them _not_ copied or selectable.  I tried playing with `user-select: none` and the `:first-letter` selector, but it's not quite doing what I want.

Putting this up for ongoing feedback as we work through this.

/cc @taniarascia , @bpas247

### CODE REVIEW

File: docs/introduction/QuickStart.mdx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+### What does Redux include?
+
+The core Redux library consists of a few API functions. The most important API is the `createStore` function.
+
+<Note type="error">
+
+**TODO** figure out what else to put here, or if this should be dropped
+
+</Note>
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]: Personally I don't think this section is needed in a quick start, since it's specifically for getting something quickly rather than overviewing the whole API, which I think the rest of the docs introduce well enough

File: docs/introduction/QuickStart.mdx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  values={[
+    { label: 'NPM', value: 'npm', },
+    { label: 'Yarn', value: 'yarn', }
+  ]
+}>
+
+<TabItem value="npm">
+
+```bash
+npm add @reduxjs/toolkit react-redux
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]:
```suggestion
npm install @reduxjs/toolkit react-redux
```

File: docs/introduction/QuickStart.mdx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+The React project starts with a single `<App>` component that shows some "hello" text on the page. Our first goal is to create a Redux store instance
+that holds some data inside, and update the `<App>` component to read that data and show it instead.
+
+To create a Redux store, start by updating the `index.js` file to import the `configureStore` function from the `@reduxjs/toolkit` package.
+
+```diff
+import React from "react";
+import ReactDOM from "react-dom";
++import {configureStore} from "@reduxjs/toolkit";
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]: I find it a little confusing to skip straight from having configureStore to having a reducer. I think this code snippet should add a minimal example call with an empty reducer function, then the next snippet can add the example reducer function.

File: docs/introduction/QuickStart.mdx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import App from "./App";
+
++function rootReducer() {
++    return "Hello Redux!";
++}
+
+
+const rootElement = document.getElementById("root");
+ReactDOM.render(<App />, rootElement);
+```
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]: I like the idea of introducing the root reducer as soon as the store is configured, but I think it might be more clear to a beginner what the reducer is used for if there's a more real-world example state here. For example, it could have `return { firstName: "Homer", lastName: "Simpson" }` and then later add actions/reducers/slices for updating the name fields, showing the relationship between the initial state and how Redux stores are updated as they're developed.

Comment (markerikson) [reply docs/introduction/QuickStart.mdx]: My intent is to have the first step simply be `() => "Hello Redux"` + `const helloText = useSelector(state => state)`, and get that to show up on screen instead of the initial text in `<App>`.  Then, once they've seen the initial connection sequence, we can move on to filling out actual app logic.

**Comment by markerikson** [comment]
Finished the "Setup and Installation" part of the page.

I'm feeling very split on the content so far.  I think what I've got here is decent, but at the same time I'm worried it's too low-level / introductory, and taking too long to get to anything "useful" and "real".

File: docs/introduction/QuickStart.mdx [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+### Prerequisites
+
+There are several things you should know in advance before learning Redux:
+
+- Basic familiarity with HTML & CSS, and the DOM.
+- Basic knowledge of JavaScript and programming.
+- Familiarity with ES6 syntax and features.
+- Basic knowledge of React, including components, props, and state.
+- Node.js and `npm` installed globally.
```
Comment (bpas247) [inline docs/introduction/QuickStart.mdx]: should we link to external resources here? In case the end user doesn't know these pre-reqs.

File: docs/introduction/QuickStart.mdx [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+If you look at the console in your browser's DevTools (or in CodeSandbox), you should see that the words `"Hello Redux!"` were printed.
+
+### Set Up the Provider Component
+
+We've got a Redux store with its state inside, but now we need to find a way to use that state inside our `<App>` component.
+
+It's _possible_ that we could move the store into a separate file, import that store into the `App.js` file, and call `store.getState()`. But, as the
+app gets bigger, we'd have to import the store into _every_ file. That's not a good idea.
```
Comment (bpas247) [inline docs/introduction/QuickStart.mdx]: I'm not sure how useful this information is to a beginner? Just comes off as more information than needed IMO

File: docs/introduction/QuickStart.mdx [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+  rootElement
+)
+```
+
+### Show the State in the App Component
+
+The last step for the initial setup is to update our `<App>` component to read the state value from the Redux store.
+
+We've already seen that the React-Redux `<Provider>` component makes the store available to any components inside. To go along with that, React-Redux
+includes a React hook called `useSelector`, which React function components can call to retrieve the state values from the store.
```
Comment (bpas247) [inline docs/introduction/QuickStart.mdx]: We should link to the React hooks API docs here. Might also be worth mentioning that React hooks is a pre-req for this page, since the end user could be new to the React hooks API as well.

File: website/sidebars.js [hunk 8]
```diff
@@ -2,6 +2,7 @@ module.exports = {
   docs: {
     Introduction: [
       'introduction/getting-started',
+      'introduction/quick-start',
```
Comment (bpas247) [inline website/sidebars.js]: The placement of this page is a bit awkward. Are we assuming that the end users are going to directly navigate to the "Quick Start" page rather than reading the "Getting Started" page first? Since both of these pages start out with the same information, and have a lot of duplicate information.

Also a bit strange that we go over installation in this page, but then the next page the user might navigate to is the "Installation" page.

File: docs/introduction/QuickStart.mdx (hunk 6 above)
Comment (markerikson) [reply docs/introduction/QuickStart.mdx]: Mmm... yes and no.  My thought was to justify why React-Redux's API is needed at all.

File: website/sidebars.js (hunk 8 above)
Comment (markerikson) [reply website/sidebars.js]: Yeah, agree that there's overlap, and that needs to be sorted out.  For now, my focus is on the actual content of this page.

File: docs/introduction/QuickStart.mdx (hunk 5 above)
Comment (markerikson) [reply docs/introduction/QuickStart.mdx]: Yep, I'm just trying to get _a_ list in place for now.

File: docs/introduction/QuickStart.mdx (hunk 6 above)
Comment (bpas247) [reply docs/introduction/QuickStart.mdx]: I guess I'm just worried about this "Quick Start" page having more content than it needs, since in my mind the goal of this page is to provide as minimal information as possibly needed to get started with Redux and React.

If you feel like this content fits that goal then we should definitely keep it 😃

File: website/package.json [hunk 9]
```diff
@@ -6,8 +6,8 @@
     "deploy": "docusaurus deploy"
   },
   "dependencies": {
-    "@docusaurus/core": "2.0.0-alpha.39",
-    "@docusaurus/preset-classic": "2.0.0-alpha.39",
+    "@docusaurus/core": "2.0.0-alpha.40",
+    "@docusaurus/preset-classic": "2.0.0-alpha.40",
```
Comment (bpas247) [inline website/package.json]: Outta curiosity, is there a specific reason that these are getting version bumps? Might be better to branch out these version bumps to a separate PR.

**Comment by DylanVann** [comment]
I think you should get the users to have some simple working code and get them interested/excited about it before going into a bunch of explanations and prerequisites and stuff.

---

### 1. Create a Create React App app:

```bash
npx create-react-app redux-counter-example
```

### 2. Install `@reduxjs/toolkit-react`

```bash
yarn add @reduxjs/toolkit-react
```

### 3. Paste this example into `index.js`

```tsx
import React from "react"
import { render } from "react-dom"
import {
  Provider
  useSelector,
  useDispatch,
  createAction as createEvent,
  configureStore as createStore,
  createReducer,
} from "@reduxjs/toolkit-react"

const onIncrementClicked = createEvent("incrementClicked")
const onDecrementClicked = createEvent("decrementClicked")

const counterReducer = createReducer(0, builder =>
  builder
    .addCase(onIncrementClicked, state => state + 1)
    .addCase(onDecrementClicked, state => state - 1),
)

const store = createStore({
  reducer: {
    counter: counterReducer,
  },
})

function App() {
  const count = useSelector(state => state.counter)
  const dispatch = useDispatch()
  return (
    <div>
      <button onClick={() => dispatch(onDecrementClicked())}>-</button>
      {count}
      <button onClick={() => dispatch(onIncrementClicked())}>+</button>
    </div>
  )
}

render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById("root"),
)
```

Open the [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) and try (there should be some gifs showing how to do this stuff):
- Clicking the buttons to see events.
- Viewing the state changes that happen on the events.
- Using time travel debugging to see earlier states of the app.

---

Basically I think "Quick Start" should be small and quick. It would be fine to have lots of links at the end. Or just a link at the end going to a more in depth tutorial after people are excited about having something working.

The idea of `@reduxjs/toolkit-react` would be to make it so people only have to install one package. It would just look like:

```tsx
export * from 'redux'
export * from 'react-redux'
export * from '@reduxjs/toolkit'
```

It would be nice to have a toggle somewhere on the page to change the code to TypeScript and change the CRA command to use the TS template.

**Comment by taniarascia** [comment]
@DylanVann In general, I agree - less is more. However, instead of "paste the all this code into index.js", I think it would be better to break it into chunks.

We'll be creating a simple counter application. First, make`index.js` file and import all dependencies.

```js
import React from "react"
import { render } from "react-dom"
import {
  Provider
  useSelector,
  useDispatch,
  createAction as createEvent,
  configureStore as createStore,
  createReducer,
} from "@reduxjs/toolkit-react"
```

Next, create two **events** - one for incrementing, one for decrementing.

```js
const onIncrementClicked = createEvent("incrementClicked")
const onDecrementClicked = createEvent("decrementClicked")
```

Then we'll create a **reducer**.

```js
const counterReducer = createReducer(0, builder =>
  builder
    .addCase(onIncrementClicked, state => state + 1)
    .addCase(onDecrementClicked, state => state - 1),
)
```

...etc.

File: docs/introduction/QuickStart.mdx [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+ReactDOM.render(<App />, rootElement)
+```
+
+In order to actually create a Redux store, we need to define our first "reducer" function.
+
+<Note type="success">
+
+#### Key Concept: "Reducers"
+
+A **_reducer_** is a function that know how to calculate a new state value when asked. A Redux store requires a single reducer function
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]:
```suggestion
A **_reducer_** is a function that knows how to calculate a new state value when asked. A Redux store requires a single reducer function
```

File: docs/introduction/QuickStart.mdx [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+when it is created, which we call the "root reducer" to say that it is the starting point for any other reducer functions inside.
+
+Every reducer is called with two arguments: the current `state`, and an object called an **`action`**. We'll look at actions a bit later.
+
+Reducers must always follow certain rules inside. As a quick summary:
+
+- Reducers must always return a state value, never `undefined`
+- Reducers can only decide what the next state value should be by looking at the `state` and `action` arguments
+- Reducers are not allowed to make changes to the `state` argument, or anything else. They can only create the new state value by making copies of the originals, and updating the copies.
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]: This might be confusing to Redux Toolkit users (and we recommend it to beginners), as mutations are supported via Immer and are recommended in the docs.

File: docs/introduction/QuickStart.mdx (hunk 6 above)
Comment (nickmccurdy) [reply docs/introduction/QuickStart.mdx]: I've seen a lot of beginners try to use vanilla Redux in React apps and they get very confused with how to handle the store. I agree with Mark here that it's important for us to explain why React Redux is helpful right away.

Of course, if we have a guide for non-React users in the future, this wouldn't be necessary.

File: docs/introduction/QuickStart.mdx (hunk 10 above)
Comment (taniarascia) [reply docs/introduction/QuickStart.mdx]: "A reducer is a function that calculates a new state value."

File: docs/introduction/QuickStart.mdx [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+  return (
+    <div className="App">
+      <h1>{helloText}</h1>
+    </div>
+  )
+}
+```
+
+We import the `useSelector` function from React-Redux, and then create a function called `selectHelloText`. It receives whatever state is in the Redux
+store as an argument. For now, we'll just return that entire state value.
```
Comment (nickmccurdy) [inline docs/introduction/QuickStart.mdx]: I've seen beginners frequently return the entire state when they don't need to in `mapStateToProps`, causing unnecessary rerenders and defeating the purpose of this API. I understand that. There's just one value in the store in this case, but I think we should at least note that only the state the component needs should be returned, as Redux optimizes rerenders based on the result.

**Comment by nickmccurdy** [comment]
Sorry I fucked up that last review, those comments are supposed to be part of my suggested changes

File: docs/introduction/QuickStart.mdx [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+
+Create a new React project sandbox by browsing to the [CodeSandbox "Create Sandbox" page](https://codesandbox.io/s/) and selecting the official React template.
+
+Once the sandbox has been created, click the "Add Dependency" button on the left, search for `@reduxjs/toolkit`, add it, and then do the same for `react-redux`.
+
+### Create a Redux Store
+
+The React project starts with a single `<App>` component that shows some "hello" text on the page. Our first goal is to create a Redux store instance
+that holds some data inside, and update the `<App>` component to read that data and show it instead.
+
```
Comment (dceddia) [inline docs/introduction/QuickStart.mdx]: I think it'd be good to show a full working React example before adding the Redux store. Maybe something with state.

Presumably, landing people on this page have used React a little. Ideally they have, anyway. Starting with a React example and "refactoring it to Redux" would:

- help a person who already knows React feel grounded, adding new knowledge on top of old
- help a person who has never seen React realize that they should go learn that first (provide a link to official tutorial) and that Redux builds on top of React

File: docs/introduction/QuickStart.mdx [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+<Note type="success">
+
+#### Key Concept: "Store"
+
+A Redux **_store_** is a special object that holds some data inside, such as a list of users that was fetched from the server.  
+The data inside the store object is called **_state_**.
+
```
Comment (dceddia) [inline docs/introduction/QuickStart.mdx]: Could add some kind of visual here to hammer home the idea that the store holds the state.

Nit below: "**Each** Redux store..." kind of implies that an app might have more than one. "A" or "Every" might be better?

File: docs/introduction/QuickStart.mdx [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+In order to actually create a Redux store, we need to define our first "reducer" function.
+
+<Note type="success">
+
+#### Key Concept: "Reducers"
+
+A **_reducer_** is a function that know how to calculate a new state value when asked. A Redux store requires a single reducer function
+when it is created, which we call the "root reducer" to say that it is the starting point for any other reducer functions inside.
```
Comment (dceddia) [inline docs/introduction/QuickStart.mdx]: Maybe skip talking about "root reducers" until later. Get one working first before revealing that there can be a tree of them.

File: docs/introduction/QuickStart.mdx [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+
+Every reducer is called with two arguments: the current `state`, and an object called an **`action`**. We'll look at actions a bit later.
```
Comment (dceddia) [inline docs/introduction/QuickStart.mdx]: Mention here that the reducer "returns the new computed state".

Could also add a few words to describe actions so it's not gnawing at them - "...an object called an **action** that describes how to update the state"

File: docs/introduction/QuickStart.mdx (hunk 11 above)
Comment (dceddia) [reply docs/introduction/QuickStart.mdx]: Yeah this is a good point. Later on, it would be good to show "here's the hard way (ignoring that immer is installed)" and "here's the easy way because of immer".

File: docs/introduction/QuickStart.mdx (hunk 6 above)
Comment (dceddia) [reply docs/introduction/QuickStart.mdx]: If this stays, it'd be good to explain _why_ "That's not a good idea." - it will get messy? hard to manage? but there's an easier way?

File: docs/introduction/QuickStart.mdx [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+The last step for the initial setup is to update our `<App>` component to read the state value from the Redux store.
+
+We've already seen that the React-Redux `<Provider>` component makes the store available to any components inside. To go along with that, React-Redux
+includes a React hook called `useSelector`, which React function components can call to retrieve the state values from the store.
+
+<Note type="success">
+
+#### Key Concept: "Selectors"
+
+A **_selector_** is a function that takes the Redux store state as an argument, and returns some part of that store state.
```
Comment (dceddia) [inline docs/introduction/QuickStart.mdx]: It quite literally "selects" a portion of the state.

**Comment by markerikson** [comment]
Haven't given up on this, but I decided I ought to take a step back and take notes on several excellent React+Redux tutorials (and some totally non-Redux-related tutorials) to get a feel for how they handle explanations first.  Then I'll try to take some inspiration from those and take another stab at this.

File: docs/introduction/QuickStart.mdx (hunk 11 above)
Comment (markerikson) [reply docs/introduction/QuickStart.mdx]: Yep, this is still my #1 concern with trying to explain use of RTK in general, and especially in this page.

**Comment by markerikson** [comment]
Should have closed this a while ago.
